### PR TITLE
ENH: Better error message for __array_ufunc__ not implemented

### PR DESCRIPTION
**Pull request by shoyer**
New behavior:

    >>> import numpy as np

    >>> class Dummy:
    ...     def __array_ufunc__(self, *args, **kwargs):
    ...         return NotImplemented

    >>> np.negative(Dummy())
    TypeError: operand type(s) do not implement __array_ufunc__(
    <ufunc 'negative'>, '__call__', <__main__.Dummy object at 0x1106df8d0>):
    'Dummy'

EDIT: updated error message to final implementation

### CODE REVIEW

**Comment by mhvk** [comment]
Ah, that `_Format` is much more powerful than I realized -- learned something new!

**Comment by shoyer** [comment]
The latest version switches up the implementation to print it's best impression of the `__array_ufunc__` method call, i.e.,
`>>> np.negative(Dummy(), out=1)`
`TypeError: operand type(s) do not implement __array_ufunc__(<ufunc 'negative'>, '__call__', <__main__.Dummy object at 0x1106df8d0>, out=(1,))`

I coded up some variations only printing the operand types (https://github.com/numpy/numpy/pull/8247#issuecomment-296346933), but I wasn't really happy with the formatting on any of those. This version has the bonus of making it completely unambiguous how the error occurred.

It uses @eric-wieser's trick to import the implementation from Python -- please check if I did that right!

File: numpy/core/src/umath/override.c [hunk 1]
```diff
@@ -429,8 +429,24 @@ PyUFunc_CheckOverride(PyUFuncObject *ufunc, char *method,
...
+            PyObject *errmsg;
+
+            if (errmsg_formatter == NULL) {
+                PyObject *mod = PyImport_ImportModule("numpy.core._internal");
+                if (mod != NULL) {
+                    errmsg_formatter = PyObject_GetAttrString(
+                        mod, "array_ufunc_errmsg_formatter");
+                    Py_DECREF(mod);
+                }
+            }
```
Comment (eric-wieser) [inline numpy/core/src/umath/override.c]: Is `npy_import_cache` available here?

**Comment by eric-wieser** [comment]
Can we distinguish somehow in the error message exactly which dispatches were tried? In particular, distinguishing "do not implement" from "implement but return `NotImplemented`" would be handy

File: numpy/core/src/umath/override.c (hunk 1 above)
Comment (shoyer) [inline numpy/core/src/umath/override.c]: Yes, done.

**Comment by shoyer** [comment]
Unit tests on my fork are passing for this branch (though @charris might consider enabling them for his fork on Travis-CI, too, so we can see checkmarks in pull requests).

> Can we distinguish somehow in the error message exactly which dispatches were tried? In particular, distinguishing "do not implement" from "implement but return NotImplemented" would be handy

I'm open to concrete suggestions, but it gets increasingly difficult to convey this nuance in an interpretable/concise error message. (Also, keeping track of this would require some small additional amount of book-keeping inside the dispatch loop.)

In part, I tried to follow the lead from Python here, which conveys even less information (it doesn't even show the input values).

Possibly a slight better version, following Python's lead, would be:
`>>> np.negative(Dummy(), out=1)`
`TypeError: operand type(s) do not implement __array_ufunc__(<ufunc 'negative'>, '__call__', ..., out=...): 'Dummy', 'object'`
where `out=...` only appears if there is an output argument.

Or perhaps we should stick with the full repr for the arguments, adding type information on at the end for increased clarity:
`TypeError: operand type(s) do not implement __array_ufunc__(<ufunc 'negative'>, '__call__', <__main__.Dummy object at 0x1106df8d0>, out=(1,)): 'Dummy', 'object'`

**Comment by shoyer** [comment]
I went ahead and added the list of type names, like how Python's built-in operations handle it.

I still have included the full call information, since I think that could be useful for debugging.

I'm not convinced it makes sense to list which dispatches were tried. Unless subclasses are involved, we look for and call `__array_ufunc__` on every argument, so the procedure for debugging this is pretty straightforward: check for `__array_ufunc__` on each argument.

**Comment by charris** [comment]
I think where and what are sufficient for most debugging. If more information is needed we will find out what is most useful after people run into problems.

**Comment by charris** [comment]
@shoyer Could you squash these? I'd do it in the merge, but I'm not sure what the proper commit message would be at this point.

**Comment by shoyer** [comment]
@charris Squashed and force-pushed. Travis-CI is testing my branch now: https://travis-ci.org/shoyer/numpy/builds/225277154

File: numpy/core/src/umath/override.c [hunk 2]
```diff
@@ -429,8 +429,19 @@ PyUFunc_CheckOverride(PyUFuncObject *ufunc, char *method,
...
             /* No acceptable override found. */
-            PyErr_SetString(PyExc_TypeError,
-                            "__array_ufunc__ not implemented for this type.");
+            static PyObject *errmsg_formatter = NULL;
+            PyObject *errmsg;
+
+            npy_cache_import("numpy.core._internal",
+                             "array_ufunc_errmsg_formatter",
+                             &errmsg_formatter);
+            errmsg = PyObject_Call(errmsg_formatter, override_args,
```
Comment (charris) [inline numpy/core/src/umath/override.c]: Note that `npy_cache_import` may fail, in which case `errmsg_formatter` will be NULL.

**Comment by charris** [comment]
You should also probably check for failed import from `_internal`, although I expect that if that fails it is the least of our problems.

**Comment by charris** [comment]
Thanks Stephan.

**Pull request merged**